SKU search in product selector

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #9150
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds the possibility to search products by SKU (partial match) in the product selector. The UX follows the iOS implementation as described here: pdfdoF-2Wl-p2

### Testing instructions
1. Create a new order and click the "Add products" button to open the product selector.
2. Verify that when the search field is active the search type options appear.
3. Verify that it's possible to search products by SKU.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img src=https://github.com/woocommerce/woocommerce-android/assets/4527432/a7473853-0661-46a0-a5dd-f377557493b1 height=600/>


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Refactor data layer to support SKU search

* Fixed selecting/unselecting variations from search results list - introduced ListItem sealed class with ProductListItem and VariationListItem child classes. Updated `onProductClick` and `Product.toUiModel` functions to handle the case of search results list containing variation items.

* Added searchType into view state

* Updated repository and list handler to accept sku search option as an argument

samiuelson committed:
Implement basic UI for search type switching

samiuelson committed:
Add unit tests

samiuelson committed:
Update release notes

samiuelson committed:
Clean up code

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModel.kt:
```diff
@@ -266,31 +293,47 @@ class ProductSelectorViewModel @Inject constructor(
-                    item.selectedVariationIds,
-                    productSelectorFlow,
-                    productSource
+        if (item is ProductListItem) {
```
Would it make sense to use `when` here? Also, extract every case to a method to give more clarity on what this piece of code doing

samiuelson replied:
Refactored in 3838fc9b07f3aaa28b7872ebf772771eba9bdbe6

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/Buttons.kt:
```diff
@@ -332,6 +368,17 @@ private fun ButtonsPreview() {
                 icon = Icons.Default.Add
             )
+
+            WCSelectableChip(
```
Shell, we make the column scrollable, so these buttons visible on the preview?

```
.verticalScroll(rememberScrollState())
```

samiuelson replied:
Done in 4f5b8d787c6620bbd17769e9d829a442d11264ac

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/Buttons.kt:
```diff
@@ -183,6 +184,41 @@ fun WCOutlinedButton(
+            leadingIcon()
+            Spacer(modifier = Modifier.width(dimensionResource(id = R.dimen.minor_100)))
+        }
+        Text(text = text)
```
The text color should be white in the dark mode?


<img width="217" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/e7f27b94-27ae-4a7e-8242-a80d5b67bfbf">
<img width="208" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/4c1b9b13-a1d4-4154-9943-9d6c9bfdb242">

samiuelson replied:
Fixed in 22e07a5d4f8ca63adc50770b4ca5e0d79f91414a

kidinov commented:
We use 2 UIs for the main product search and here. Are there any considerations on merging or aligning them?

<img width="401" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/6530cb62-1ad6-41af-ae8e-96ba365d0414">

kidinov commented:
When we are searching for something, then we start to use filters - we show a screen "that we can't find anything with search request for some time while filters being applied. Probably we should show loading state

https://github.com/woocommerce/woocommerce-android/assets/4923871/bb0d2df2-9019-43b5-8bc9-d1fc19f0a894

kidinov commented:
Search text in the dark mode probably should be white - it is white on other search screens

<img width="194" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/769572e8-245c-4d0e-a024-9b6271c84400">

kidinov commented:
Shell we change the back button to "back" when we are in search mode? Now I find myself clicking this button expecting an exit from the search, but it leaves the screen completely

Also, it will make clear UX wise that we do not not offer at the moment text search with filters applied (maybe even hide filtering options when user in the search mode)

<img width="117" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/a63097f3-1cc6-44d0-9194-87bd02860fd8">

<img width="232" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/bd026e45-a60b-459b-a487-f916a4b678f8">

kidinov commented:
After activity death we don't recover search state

<img width="346" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/06f4d8e8-438b-4bd2-9093-a70c0226d2d9">

kidinov commented:
`"` on a new line?

<img width="343" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/8e6f3c4a-a24a-49d2-a976-8d6fda815a4b">

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorScreen.kt:
```diff
@@ -119,6 +116,54 @@ fun ProductSelectorScreen(
+                        .padding(dimensionResource(id = dimen.minor_100))
+                        .weight(1f),
+                    onClick = { onSearchTypeChanged(SearchType.DEFAULT) },
+                    text = stringResource(id = string.product_search_all),
```
I do not entirely understand the naming here. When as a user I select "SKU" I understand that my text is matched by the SKU of a product. What do we match by when we say to the user "all products"?

samiuelson replied:
Yeah, I agree, the "All products" label is unclear. Maybe it could be renamed to "Quick search"? My suggestion is to follow now the current convention and make it consistent with iOS and the Products screen search. Next, let's raise a discussion on adjusting the name or redesigning the UX. Wdut?

samiuelson committed:
Make preview's column scrollable

samiuelson committed:
Fix search state restoring after activity death

samiuelson committed:
Hide Filter button when search is active

Search is hidden in case of active filters as well. This way we avoid confusion - filters ignore search results and when active are applied to the whole product list. Analogically, the search results ignore filters.

kidinov reviewed (commented):
Good job @samiuelson! I left some comments

This is a complex change, so I think we may benefit if @AnirudhBhat would test this as well

samiuelson committed:
Clear search query on back press before going back

* Implemented BackPressListener by ProductSelectorFragment
* Cleaned up view's search state model
* Clearing focus on back press if search is active
* Requesting focus when returning back to activity after previously active search

samiuelson committed:
Clean up code

samiuelson commented:
> When we are searching for something, then we start to use filters - we show a screen "that we can't find anything with search request for some time while filters being applied. Probably we should show loading state
> 

Good catch! Fixed with 32191afd580bf1e85edd0982c9363d4ca5cd1b94. 
When search results are shown, Filter button is hidden now. Analogically, when filters are active we hide the search field. This way we don't confuse the users because filters are not applied to the search results but to all products. The same with search - we search through all the products, not only the filtered ones.

samiuelson commented:
> After activity death we don't recover search state
> 
Fixed in f2bd1d0bf763f69f5c8e878dee18a37d71f377df

samiuelson commented:
> Shell we change the back button to "back" when we are in search mode? Now I find myself clicking this button expecting an exit from the search, but it leaves the screen completely
> 
> Also, it will make clear UX wise that we do not not offer at the moment text search with filters applied (maybe even hide filtering options when user in the search mode)
> 
> <img alt="image" width="117" src="https://user-images.githubusercontent.com/4923871/242530913-a63097f3-1cc6-44d0-9194-87bd02860fd8.png"> <img alt="image" width="232" src="https://user-images.githubusercontent.com/4923871/242531141-bd026e45-a60b-459b-a487-f916a4b678f8.png">

Fixed in 8e5e843ed2a1424ffa465fca84ffae72dfa8e8ff. Clicking the navigation icon on the toolbar and the device's back button will now reset the search first before closing the screen. Result:

https://github.com/woocommerce/woocommerce-android/assets/4527432/ab197e49-3b55-49eb-be69-27cb69e8eca7

I agree it would be nice to show "back" icon instead of "close" icon in case the search is active. To do that we'll need to hide MainActivity's Toolbar and migrate to Compose Scaffold with Toolbar. I'd like to do it as an improvement in the next PR.

samiuelson committed:
Update WCSearchField's text and cursor colors

samiuelson commented:
> Search text in the dark mode probably should be white - it is white on other search screens

Updated in f39ceac7331ee5d196e4401db9ee70c80a672688

samiuelson committed:
Update WCSelectableChip content colors

samiuelson commented:
> `"` on a new line?
> 
> <img alt="image" width="343" src="https://user-images.githubusercontent.com/4923871/242532547-8e6f3c4a-a24a-49d2-a976-8d6fda815a4b.png">

I documented this behavior with the issue: https://github.com/woocommerce/woocommerce-android/issues/9169

It's unlikely to happen because it's not possible to reproduce it using the device's keyboard. I guess it can happen when pasting copied text or using an external keyboard. In such cases, we should probably remove line separator chars from the search query. IMO, in real life, BasicTextField's `singleLine = true` argument should be sufficient. WDUT?

samiuelson commented:
> We use 2 UIs for the main product search and here. Are there any considerations on merging or aligning them?
> 
I agree it would be nice to reuse Product Selector in the main Products screen and remove a bunch of the View-based code around `ProductListFragment`. However, I think it would be hard to justify such refactoring effort just to clean up the code. So probably any merge won't happen until the Products screen becomes problematic to maintain.

samiuelson committed:
Fix instrumented tests

samiuelson committed:
Remove unused code

samiuelson committed:
Simplify the code

samiuelson commented:
> Good job @samiuelson! I left some comments
> 
> This is a complex change, so I think we may benefit if @AnirudhBhat would test this as well

Thanks for the thorough review @kidinov! I addressed your comments.

@AnirudhBhat I'd appreciate your additional help with testing this feature if you have the capacity.

AnirudhBhat commented:
>I'd appreciate your additional help with testing this feature if you have the capacity.

@samiuelson On it. Thanks for the heads up.

kidinov commented:
@samiuelson Thank you for all the improvements!

The only undesired behavior I noticed now is that we show an empty state after we select a filter. Shell we show a loading state here as well?

https://github.com/woocommerce/woocommerce-android/assets/4923871/73bf4ed3-9196-48f1-86af-d6903f82327e

AnirudhBhat reviewed (approved):
Thanks for working on this @samiuelson. The functionality works as expected. As @kidinov  mentioned, I can also see the empty state for some time before the filters are applied. 

Not a blocking comment - Also, seems like there are some significant changes in the UI code. We could probably write some compose UI tests for those in a separate PR.

samiuelson committed:
Improve UI of variable subscription items

samiuelson committed:
Fix detekt issues
